File: Go-SDL2/message.go
package main

import (
	"github.com/veandco/go-sdl2/sdl"
	"github.com/veandco/go-sdl2/ttf"
)

type message struct {
	_renderer  *sdl.Renderer
	_font      *ttf.Font
	_fontColor sdl.Color
	_image     *sdl.Texture
	_rect      sdl.Rect
}

func newMessage(renderer *sdl.Renderer) (*message, error) {
	m := &message{}
	var err error

	m._renderer = renderer

	if m._font, err = ttf.OpenFont("fonts/freesansbold.ttf", messageFontSize); err != nil {
		return m, err
	}

	m._fontColor = sdl.Color{R: 255, G: 255, B: 255, A: 255}

	return m, err
}

func (m *message) close() {
	m._font.Close()
	m._image.Destroy()
}

func (m *message) message(text string) error {
	var messageSurf *sdl.Surface
	var err error

	if messageSurf, err = m._font.RenderUTF8Blended(text, m._fontColor); err != nil {
		return err
	}
	defer messageSurf.Free()

	m._image.Destroy()
	if m._image, err = m._renderer.CreateTextureFromSurface(messageSurf); err != nil {
		return err
	}

	if _, _, m._rect.W, m._rect.H, err = m._image.Query(); err != nil {
		return err
	}

	m._rect.X = (windowWidth - m._rect.W) / 2
	m._rect.Y = (windowHeight - m._rect.H) / 2

	return err
}

func (m *message) draw() {
	m._renderer.Copy(m._image, nil, &m._rect)
}

File: C-SDL2/message.c
#include "message.h"

bool message_new(struct Message **message, SDL_Renderer *renderer) {
    *message = calloc(1, sizeof(struct Message));
    if (!*message) {
        fprintf(stderr, "Error in calloc of message!\n");
        return true;
    }
    struct Message *m = *message;

    m->renderer = renderer;
    m->color = (SDL_Color){255, 255, 255, 255};

    m->font = TTF_OpenFont("fonts/freesansbold.ttf", MESSAGE_FONT_SIZE);
    if (!m->font) {
        fprintf(stderr, "Error creating font: %s\n", TTF_GetError());
        return true;
    }

    return false;
}

void message_free(struct Message **message) {
    if (*message) {
        TTF_CloseFont((*message)->font);
        (*message)->font = NULL;
        SDL_DestroyTexture((*message)->image);
        (*message)->image = NULL;
        (*message)->renderer = NULL;
        free(*message);
        *message = NULL;
    }
}

bool message_update_level(struct Message *m, Uint32 level) {
    int length = snprintf(NULL, 0, "Level: %d", level) + 1;
    char level_str[length];
    snprintf(level_str, (size_t)length, "Level: %d", level);
    if (message_update(m, level_str)) {
        return true;
    }
    return false;
}

bool message_update(struct Message *m, const char* text) {
    if (m->image) {
        SDL_DestroyTexture(m->image);
    }

    SDL_Surface *surface = TTF_RenderText_Blended(m->font, text, m->color);
    if (!surface) {
        fprintf(stderr, "Error creating a surface: %s\n", SDL_GetError());
        return true;
    }

    m->image = SDL_CreateTextureFromSurface(m->renderer, surface);
    if (!m->image) {
        fprintf(stderr, "Error creating a texture: %s\n", SDL_GetError());
        return true;
    }

    SDL_FreeSurface(surface);

    if (SDL_QueryTexture(m->image, NULL, NULL, &m->rect.w, &m->rect.h)) {
        fprintf(stderr, "Error while querying texture: %s\n", SDL_GetError());
        return true;
    }

    m->rect.x = (WINDOW_WIDTH - m->rect.w) / 2;
    m->rect.y = (WINDOW_HEIGHT - m->rect.h) /2 ;
    
    return false;
}

bool message_draw(struct Message *m) {
    if (SDL_RenderCopy(m->renderer, m->image, NULL, &m->rect)) {
        fprintf(stderr, "Error while rendering texture: %s\n", SDL_GetError());
        return true;
    }

    return false;
}
